File: python/examples/chisquare.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function, division

# Standard library stuff
import re
import warnings
from time import time
from contextlib import contextmanager

# GooFit package
from goofit import *

import numpy as np

# Protect the matplotlib call for systems with no graphics
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt


try:
    import uncertainties as unc
    from uncertainties import unumpy
except ImportError:
    unc = None
    warnings.warn("uncertainties not found, no unc on plot", RuntimeWarning)

# Optional import of numba, if available
# Make code much faster if found!
try:
    from numba import jit
except ImportError:
    def jit(*args, **kargs):
        def copyf(function):
            return function
        return copyf
    warnings.warn("Numba not found, will be 100x slower. Try `pip install numba` to install.", RuntimeWarning)

# Simple timer in a context manager
@contextmanager
def timed(msg):
    start = time()
    try:
        yield
    finally:
        end = time()
        print(msg, end - start, "s")

@jit(nopython=True)
def integralExpCon(lo, hi):
    return(np.exp(-lo) - np.exp(-hi))

@jit(nopython=True)
def integralExpLin(lo, hi):
    return((lo + 1) * np.exp(-lo) - (hi + 1) * np.exp(-hi))

@jit(nopython=True)
def integralExpSqu(lo, hi):
    return((lo * lo + 2 * lo + 2) * np.exp(-lo) - (hi * hi + 2 * hi + 2) * np.exp(-hi))

@jit(nopython=True)
def generateEvents(numbins, lowerlimit, upperlimit, conCoef, linCoef, squCoef, eventsToGenerate):

    totalRSintegral = integralExpCon(0, 100)
    step            = (upperlimit - lowerlimit) / numbins

    rsEvtVec = np.empty(numbins)
    wsEvtVec = np.empty(numbins)

    for i in range(numbins):

        binStart = i * step
        binStart += lowerlimit
        binFinal = binStart + step

        rsIntegral = integralExpCon(binStart, binFinal)
        wsIntegral = conCoef * integralExpCon(binStart, binFinal)
        wsIntegral += linCoef * integralExpLin(binStart, binFinal)
        wsIntegral += squCoef * integralExpSqu(binStart, binFinal)

        expectedRSevts = eventsToGenerate * rsIntegral / totalRSintegral
        expectedWSevts = eventsToGenerate * wsIntegral / totalRSintegral

        rsEvtVec[i] = np.random.poisson(expectedRSevts)
        wsEvtVec[i] = np.random.poisson(expectedWSevts)

    return rsEvtVec, wsEvtVec


def produce_hist(rsEvts, wsEvts):

    rsEvts[rsEvts == 0] = 1 # Cheating to avoid div by zero.
    ratio = wsEvts / rsEvts
    wsEvts[wsEvts == 0] = 1 # Avoid zero errors

    error = np.sqrt(wsEvts / rsEvts**2 +  wsEvts**2 / rsEvts**3)
    return ratio, error


def fitRatio(decayTime, weights, rsEvts, wsEvts, plotname=None):

    ratioData = BinnedDataSet(decayTime)

    ratio, error = produce_hist(rsEvts, wsEvts)

    for i in range(len(ratio)):
        ratioData.setBinContent(i, ratio[i])
        ratioData.setBinError(i, error[i])

    poly = PolynomialPdf("poly", decayTime, weights)
    poly.setFitControl(BinnedErrorFit())
    poly.setData(ratioData)
    fitter = FitManager(poly)
    fitter.verbosity = 0
    mn_func_min = fitter.fit()
    mn_cov = mn_func_min.UserCovariance()
    cov = mn_cov.to_matrix()
    print(cov)

    xvals = np.array(decayTime.bin_centers)
    p = [w.value for w in reversed(weights)]
    print("GooFit fit:", end=' ')
    plot_ratio(xvals, ratio, error, p, plotname, cov)

    # This would work just fine, too
    # values = poly.evaluateAtPoints(decayTime)
    # plot.plot(xvals, values)

    return fitter


def fitRatioCPU(decayTime, rsEvts, wsEvts, plotname=None):
    ratio, error = produce_hist(rsEvts, wsEvts)
    xvals = np.array(decayTime.bin_centers)

    p, cov = np.polyfit(xvals, ratio, 2, w=1/error, cov=True)

    print("CPU Fit:", end=' ')
    plot_ratio(xvals, ratio, error, p, plotname, cov)

def plot_ratio(xvals, ratio, error, p, plotname, cov=None):
    result = "{:.2e} t^2 + {:.2e} t + {:.2e}".format(*p)
    result = result.replace('e-0','e-')
    print("Fit result:", result)

    if plotname:
        result = re.sub(r'e(\S+?)(\s|$)', r' \\times {10}^{\1} ', result)
        fig, ax = plt.subplots()
        ax.errorbar(xvals, ratio, yerr=error, fmt='k.')
        ax.plot(xvals, np.polyval(p, xvals))
        ax.text(.1, .85, "${}$".format(result),
                transform=ax.transAxes, fontsize=16)

        if cov is not None and unc is not None:

            p_unc = unc.correlated_values(p, cov)
            y = np.polyval(p_unc, xvals)
            ax.fill_between(xvals,
                     unumpy.nominal_values(y)-unumpy.std_devs(y),
                     unumpy.nominal_values(y)+unumpy.std_devs(y),
                     alpha=.5, facecolor='b')

        fig.savefig(plotname)


def main():
    decayTime = Observable("decayTime", 0, 10)

    numbins = 100

    decayTime.value = 100
    decayTime.numbins = numbins

    rSubD = 0.03
    rBarD = 0.03
    delta = 0
    wpPhi = 0
    x_mix = 0.0016
    y_mix = 0.0055
    magPQ = 1.0
    magQP = 1.0 / magPQ

    eventsToGenerate = 10000000

    dZeroLinearCoef = magPQ * np.sqrt(rSubD) * (y_mix * np.cos(delta + wpPhi) - x_mix * np.sin(delta + wpPhi))
    d0barLinearCoef = magQP * np.sqrt(rBarD) * (y_mix * np.cos(delta - wpPhi) - x_mix * np.sin(delta - wpPhi))

    dZeroSecondCoef = 0.25 * magPQ * magPQ * (x_mix * x_mix + y_mix * y_mix)
    d0barSecondCoef = 0.25 * magQP * magQP * (x_mix * x_mix + y_mix * y_mix)

    dZeroEvtsRS, dZeroEvtsWS = generateEvents(
            decayTime.numbins, decayTime.lowerlimit, decayTime.upperlimit,
            rSubD, dZeroLinearCoef, dZeroSecondCoef, eventsToGenerate)

    d0barEvtsRS, d0barEvtsWS = generateEvents(
            decayTime.numbins, decayTime.lowerlimit, decayTime.upperlimit,
            rBarD, d0barLinearCoef, d0barSecondCoef, eventsToGenerate)

    constaCoef = Variable("constaCoef", 0.03, 0.01, -1, 1)
    linearCoef = Variable("linearCoef", 0, 0.01, -1, 1)
    secondCoef = Variable("secondCoef", 0, 0.01, -1, 1)

    weights = (constaCoef, linearCoef, secondCoef)

    retval1 = fitRatio(decayTime, weights, dZeroEvtsRS, dZeroEvtsWS, "chisquare_dzeroEvtRatio_goo_python.png")
    retval2 = fitRatio(decayTime, weights, d0barEvtsRS, d0barEvtsWS, "chisquare_dzbarEvtRatio_goo_python.png")

    with timed("Total CPU (2x fits)"):
        fitRatioCPU(decayTime, dZeroEvtsRS, dZeroEvtsWS, "chisquare_dzeroEvtRatio_cpu_python.png")
        fitRatioCPU(decayTime, d0barEvtsRS, d0barEvtsWS, "chisquare_dzbarEvtRatio_cpu_python.png")

    print("Exit codes (should be 0):", int(retval1), "and", int(retval2))

    return (int(retval1) + int(retval2))



if __name__ == "__main__":
    # Print info about your system and GooFit
    print_goofit_info()

    # Run the main program
    main()

File: examples/chisquare/chisquare.cpp
#include <goofit/Application.h>
#include <goofit/BinnedDataSet.h>
#include <goofit/FitControl.h>
#include <goofit/FitManager.h>
#include <goofit/UnbinnedDataSet.h>
#include <goofit/Variable.h>

#include <CLI/Timer.hpp>

#include <TCanvas.h>
#include <TH1F.h>
#include <TLatex.h>
#include <TMinuit.h>
#include <TRandom.h>
#include <goofit/PDFs/basic/PolynomialPdf.h>

#include <iostream>
#include <string>
#include <tuple>
#include <vector>

TCanvas foo;

using namespace std;
using namespace GooFit;

vector<double> ratios;
vector<double> errors;

// Global needed to put it in the fit function for ROOT
Observable decayTime{"decayTime", 0, 10};

double integralExpCon(double lo, double hi) { return (exp(-lo) - exp(-hi)); }

double integralExpLin(double lo, double hi) { return ((lo + 1) * exp(-lo) - (hi + 1) * exp(-hi)); }

double integralExpSqu(double lo, double hi) {
    return ((lo * lo + 2 * lo + 2) * exp(-lo) - (hi * hi + 2 * hi + 2) * exp(-hi));
}

void generateEvents(Observable decayTime,
                    vector<int> &rsEvtVec,
                    vector<int> &wsEvtVec,
                    double conCoef,
                    double linCoef,
                    double squCoef,
                    int eventsToGenerate) {
    static TRandom donram(24);
    double totalRSintegral = integralExpCon(0, 100);
    double step            = (decayTime.getUpperLimit() - decayTime.getLowerLimit()) / decayTime.getNumBins();

    for(int i = 0; i < decayTime.getNumBins(); ++i) {
        double binStart = i * step;
        binStart += decayTime.getLowerLimit();
        double binFinal = binStart + step;

        double rsIntegral = integralExpCon(binStart, binFinal);
        double wsIntegral = conCoef * integralExpCon(binStart, binFinal);
        wsIntegral += linCoef * integralExpLin(binStart, binFinal);
        wsIntegral += squCoef * integralExpSqu(binStart, binFinal);

        double expectedRSevts = eventsToGenerate * rsIntegral / totalRSintegral;
        double expectedWSevts = eventsToGenerate * wsIntegral / totalRSintegral;

        int rsEvts  = donram.Poisson(expectedRSevts);
        int wsEvts  = donram.Poisson(expectedWSevts);
        rsEvtVec[i] = rsEvts;
        wsEvtVec[i] = wsEvts;

        if(0 == (i % 10))
            std::cout << "Events in bin " << i << " : " << rsEvts << " (" << expectedRSevts << ") " << wsEvts << " ("
                      << expectedWSevts << ")\n";
    }
}

std::tuple<int, std::string> fitRatio(Observable decayTime,
                                      vector<Variable> weights,
                                      vector<int> &rsEvts,
                                      vector<int> &wsEvts,
                                      std::string plotName = "") {
    TH1D ratioHist("ratioHist", "", decayTime.getNumBins(), decayTime.getLowerLimit(), decayTime.getUpperLimit());

    BinnedDataSet ratioData(decayTime);

    for(unsigned int i = 0; i < wsEvts.size(); ++i) {
        double ratio = wsEvts[i];

        if(0 == rsEvts[i])
            rsEvts[i] = 1; // Cheating to avoid div by zero.

        ratio /= rsEvts[i];

        if(0 == wsEvts[i])
            wsEvts[i] = 1; // Avoid zero errors

        double error = wsEvts[i] / pow(rsEvts[i], 2);
        error += pow(wsEvts[i], 2) / pow(rsEvts[i], 3);
        error = sqrt(error);

        ratioData.setBinContent(i, ratio);
        ratioData.setBinError(i, error);
        ratioHist.SetBinContent(i + 1, ratio);
        ratioHist.SetBinError(i + 1, error);
    }

    PolynomialPdf poly("poly", decayTime, weights);
    poly.setFitControl(std::make_shared<BinnedErrorFit>());
    poly.setData(&ratioData);
    FitManager fitter{&poly};

    CLI::Timer timer_cpu{"GPU"};
    fitter.fit();
    std::string timer_str = timer_cpu.to_string();

    if(!plotName.empty()) {
        vector<fptype> values = poly.evaluateAtPoints(decayTime);
        TH1D pdfHist("pdfHist", "", decayTime.getNumBins(), decayTime.getLowerLimit(), decayTime.getUpperLimit());

        for(int i = 0; i < values.size(); ++i) {
            pdfHist.SetBinContent(i + 1, values[i]);
        }

        ratioHist.SetMarkerStyle(8);
        ratioHist.SetMarkerSize(0.5);
        ratioHist.SetStats(false);
        ratioHist.Draw("p");

        TString str1 = fmt::format(
            "Constant [10^{{-2}}] : {:.3} #pm {:.3}", weights[0].getValue() * 1e2, weights[0].getError() * 1e2);
        TLatex res1(0.14, 0.83, str1);
        res1.SetNDC(true);

        TString str2 = fmt::format(
            "Linear [10^{{-4}}] : {:.3} #pm {:.3}", weights[1].getValue() * 1e4, weights[1].getError() * 1e4);
        TLatex res2(0.14, 0.73, str2);
        res2.SetNDC(true);

        TString str3 = fmt::format(
            "Quadratic [10^{{-6}}]: {:.3} #pm {:.3}", weights[2].getValue() * 1e6, weights[2].getError() * 1e6);
        TLatex res3(0.14, 0.63, str3);
        res3.SetNDC(true);

        res1.Draw();
        res2.Draw();
        res3.Draw();

        pdfHist.SetLineColor(kBlue);
        pdfHist.SetLineWidth(3);
        pdfHist.SetStats(false);
        pdfHist.Draw("lsame");
        foo.SaveAs(plotName.c_str());
    }

    std::cout << "Polynomial function: " << poly.getCoefficient(2) << " * t^2 + " << poly.getCoefficient(1) << " * t + "
              << poly.getCoefficient(0) << std::endl;

    return make_tuple(int(fitter), timer_str);
}

void cpvFitFcn(int &npar, double *gin, double &fun, double *fp, int iflag) {
    double conCoef = fp[0];
    double linCoef = fp[1];
    double squCoef = fp[2];

    double chisq = 0;
    double step  = (decayTime.getUpperLimit() - decayTime.getLowerLimit()) / decayTime.getNumBins();

    for(unsigned int i = 0; i < ratios.size(); ++i) {
        double currDTime = decayTime.getLowerLimit() + (i + 0.5) * step;
        double pdfval    = conCoef + linCoef * currDTime + squCoef * currDTime * currDTime;
        chisq += pow((pdfval - ratios[i]) / errors[i], 2);
    }

    fun = chisq;
}

void fitRatioCPU(Observable decayTime, vector<int> &rsEvts, vector<int> &wsEvts) {
    TH1D *ratioHist
        = new TH1D("ratioHist", "", decayTime.getNumBins(), decayTime.getLowerLimit(), decayTime.getUpperLimit());

    ratios.resize(wsEvts.size());
    errors.resize(wsEvts.size());

    for(unsigned int i = 0; i < wsEvts.size(); ++i) {
        if(0 == rsEvts[i])
            rsEvts[i] = 1; // Cheating to avoid div by zero.

        fptype ratio = wsEvts[i] / rsEvts[i];

        if(0 == wsEvts[i])
            wsEvts[i] = 1; // Avoid zero errors

        double error = wsEvts[i] / pow(rsEvts[i], 2);
        error += pow(wsEvts[i], 2) / pow(rsEvts[i], 3);
        error = sqrt(error);

        ratios[i] = ratio;
        errors[i] = error;
        ratioHist->SetBinContent(i + 1, ratio);
        ratioHist->SetBinError(i + 1, error);
    }

    TMinuit *minuit = new TMinuit(3);
    minuit->DefineParameter(0, "constaCoef", 0.03, 0.01, -1, 1);
    minuit->DefineParameter(1, "linearCoef", 0, 0.01, -1, 1);
    minuit->DefineParameter(2, "secondCoef", 0, 0.01, -1, 1);
    minuit->SetFCN(cpvFitFcn);

    minuit->Migrad();
}

int main(int argc, char **argv) {
    GooFit::Application app("Chi-square example", argc, argv);

    int numbins = 100;
    app.add_option("-n,--numbins", numbins, "Number of bins", true);

    int eventsToGenerate = 10000000;
    app.add_option("-e,--events", eventsToGenerate, "Events to generate", true);

    GOOFIT_PARSE(app);

    // Time is in units of lifetime

    decayTime.setValue(100.);
    decayTime.setNumBins(numbins);

    double rSubD = 0.03;
    double rBarD = 0.03;
    double delta = 0;
    double wpPhi = 0;
    double x_mix = 0.0016;
    double y_mix = 0.0055;
    double magPQ = 1.0;
    double magQP = 1.0 / magPQ;

    vector<int> dZeroEvtsWS(decayTime.getNumBins());
    vector<int> dZeroEvtsRS(decayTime.getNumBins());
    vector<int> d0barEvtsWS(decayTime.getNumBins());
    vector<int> d0barEvtsRS(decayTime.getNumBins());

    double dZeroLinearCoef = magPQ * sqrt(rSubD) * (y_mix * cos(delta + wpPhi) - x_mix * sin(delta + wpPhi));
    double d0barLinearCoef = magQP * sqrt(rBarD) * (y_mix * cos(delta - wpPhi) - x_mix * sin(delta - wpPhi));

    double dZeroSecondCoef = 0.25 * magPQ * magPQ * (x_mix * x_mix + y_mix * y_mix);
    double d0barSecondCoef = 0.25 * magQP * magQP * (x_mix * x_mix + y_mix * y_mix);

    generateEvents(decayTime, dZeroEvtsRS, dZeroEvtsWS, rSubD, dZeroLinearCoef, dZeroSecondCoef, eventsToGenerate);
    generateEvents(decayTime, d0barEvtsRS, d0barEvtsWS, rBarD, d0barLinearCoef, d0barSecondCoef, eventsToGenerate);

    Variable constaCoef("constaCoef", 0.03, 0.01, -1, 1);
    Variable linearCoef("linearCoef", 0, 0.01, -1, 1);
    Variable secondCoef("secondCoef", 0, 0.01, -1, 1);

    vector<Variable> weights = {constaCoef, linearCoef, secondCoef};

    int retval1, retval2;
    std::string fit1, fit2;
    std::tie(retval1, fit1)
        = fitRatio(decayTime, weights, dZeroEvtsRS, dZeroEvtsWS, "chisquare_dzeroEvtRatio_goo_cpp.png");
    std::tie(retval2, fit2)
        = fitRatio(decayTime, weights, d0barEvtsRS, d0barEvtsWS, "chisquare_dzbarEvtRatio_goo_cpp.png");

    CLI::Timer timer_cpu{"Total CPU (2x fits)"};
    fitRatioCPU(decayTime, dZeroEvtsRS, dZeroEvtsWS);
    fitRatioCPU(decayTime, d0barEvtsRS, d0barEvtsWS);
    std::string cpu_string = timer_cpu.to_string();

    std::cout << fit1 << "\n" << fit2 << "\n" << cpu_string << std::endl;

    fmt::print("Exit codes (should be 0): {} and {}\n", retval1, retval2);

    return retval1 + retval2;
}
